propagation-stackdriver/src/main/java/zipkin2/propagation/stackdriver/StackdriverTracePropagation.java

```diff
@@ -0,0 +1,146 @@
+/**
+ * Copyright 2016-2018 The OpenZipkin Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package zipkin2.propagation.stackdriver;
+
+import brave.propagation.B3Propagation;
+import brave.propagation.Propagation;
+import brave.propagation.TraceContext;
+import brave.propagation.TraceContextOrSamplingFlags;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import static brave.internal.HexCodec.lenientLowerHexToUnsignedLong;
+
+public final class StackdriverTracePropagation<K> implements Propagation<K> {
+
+	public static final Propagation.Factory FACTORY = new Propagation.Factory() {
+		@Override public <K> Propagation<K> create(KeyFactory<K> keyFactory) {
+			return new StackdriverTracePropagation<>(keyFactory);
+		}
+
+		@Override public boolean supportsJoin() {
+			return true;
+		}
+
+		@Override public String toString() {
+			return "StackdriverTracePropagationFactory";
+		}
+	};
+
+	/**
+	 * 128 or 64-bit trace ID lower-hex encoded into 32 or 16 characters (required)
+	 */
+	private static final String TRACE_ID_NAME = "x-cloud-trace-context";
+
+	private static final Logger LOG = Logger.getLogger(StackdriverTracePropagation.class.getName());
+
+	private Propagation<K> b3Propagation;
+	private final K traceIdKey;
+	private List<K> fields;
+
+	StackdriverTracePropagation(KeyFactory<K> keyFactory) {
+		this.traceIdKey = keyFactory.create(TRACE_ID_NAME);
+		this.fields = Collections.unmodifiableList(Collections.singletonList(traceIdKey));
+		this.b3Propagation = B3Propagation.FACTORY.create(keyFactory);
+	}
+
+	@Override public List<K> keys() {
+		return fields;
+	}
+
+	@Override public <C> TraceContext.Injector<C> injector(Setter<C, K> setter) {
+		return b3Propagation.injector(setter);
+	}
+
+	@Override public <C> TraceContext.Extractor<C> extractor(Getter<C, K> getter) {
+		if (getter == null) throw new NullPointerException("getter == null");
+		return new CompositeExtractor<>(
+				new XCloudTraceContextExtractor<>(this, getter),
+				b3Propagation.extractor(getter));
+	}
+
+	static final class XCloudTraceContextExtractor<C, K> implements TraceContext.Extractor<C> {
```

Comment:
ok i see. maybe top-level this file and make it public? the "stackdriver" one can mention in docs that it is a composite helper and not to be used inside another composite.